splicemachine/spliceengine
# DB-12158 Automatic view refreshing for alter table add/drop column

## Short Description
This change introduces a new automatic view refreshing mode for alter table add/drop column.

3.1 PR: https://github.com/splicemachine/spliceengine/pull/5660.
3.0 PR: https://github.com/splicemachine/spliceengine/pull/5662.

## Long Description

This change introduces a new automatic view refreshing mode for alter table add/drop column.

When this mode is turned on, an `ALTER TABLE T ADD/DROP COLUMN ...` statement automatically refreshes all views defined both directly and indirectly on `T` whose top-level select list is a `SELECT *`. Other views are not refreshed.

This mode is off by default. To turn it on, run the following:
```
call syscs_util.syscs_set_global_database_property('splice.execution.alterTable.autoViewRefreshing', true)
```

To turn it off, run the following:
```
call syscs_util.syscs_set_global_database_property('splice.execution.alterTable.autoViewRefreshing', null)
```

### Detailed behavior description

- A view `V` defined using a top-level explicit select column list, i.e., `select a, b, ...` on top of `T`
	- this kind of views are not refreshed no matter view refreshing mode is on or off, even if it has subqueries that has `select *` (see explanation 1)
	- add column: the new column is not shown in `select * from v`
	- drop a column that is added after the `V` is defined: no effect
	- drop a column that exists when `V` is defined: throw (previous behavior: assertion failure if column number mismatch or wrong column names if column numbers match but it's a different set of columns)
- A view `V` defined using a top-level `select *` on top of `T`
	- when view refreshing mode is off, it has the same behavior as the explicit select column case
	- when view refreshing mode is on
		- add column: new column is shown in `select * from v`
		- drop column: dropped column is removed from `select * from v`
- A view `VV` defined using a top-level `select *` on another view `V` on top of `T`
	- when auto view refreshing mode is on, `VV` gets refreshed when a column is added to or dropped from `T`
	- if `V` is not a `select *` view, refreshing `VV` has no effect

### Explanation 1

We only need to refresh views defined with top-level "select *". View definitions could be arbitrarily complex, having "select *" in a subquery. However, as long as the top-level select items are explicitly fixed (i.e., "select a, b, ..."), newly added column will not be selected even if we refresh the view. Also, if one of the explicitly selected column is dropped, the view definition becomes invalid even if we refresh the view. As a result, in both cases, there is no need to refresh the view if its top-level select columns are explicit.

## How to test
It would be too long to copy all test cases here. For references, please check the following ITs in `DropAddColumnIT`:
- testAutomaticViewRefreshingOn_SelectStarView
- testAutomaticViewRefreshingOn_SecondLevelSelectStarView
- testAutomaticViewRefreshingOn_SelectExplicitColumnsView
- testAutomaticViewRefreshingOff
